### ISSUE TITLE
[JIT legacy executor] device propagation regression

### ISSUE THREAD

**Issue by jjsjann123** [pytorch/pytorch#34361]
## 🐛 Bug

PR #33705 introduces regression on shape propagation.
`rand_like` does not propagate correct device from input to output any more.I'm observing gpu tensor input with output marked as cpu tensor.

## To Reproduce

```
import torch
torch._C._jit_set_profiling_executor(False)
torch._C._jit_set_profiling_mode(False)

def fn(x : torch.Tensor):                  
    out = torch.nn.functional.dropout(x, p=0.1, training=True)  
    return out

fn_jit = torch.jit.script(fn)
x = torch.randn(20, 64, device="cuda", dtype=torch.float32).requires_grad_()
print(fn_jit.graph_for(x))
```

After the commit, output becomes:
```
graph(%x.1 : Float(*, *)):
  %17 : Float(*, *) = prim::DifferentiableGraph_0(%x.1)
  return (%17)
with prim::DifferentiableGraph_0 = graph(%0 : Float(*, *)):
  %51 : float = prim::Constant[value=1.1111111111111112]()
  %50 : float = prim::Constant[value=0.90000000000000002]()
  %49 : bool = prim::Constant[value=1]()
  %7 : int = prim::Constant[value=1]() # <string>:103:64
  %6 : float = prim::Constant[value=1.]() # <string>:100:18
  %5 : None = prim::Constant()
  %15 : Double(*, *) = aten::rand_like(%0, %5, %5, %5, %5, %7) # <string>:103:27
  %mask.1 : Bool(*, *) = aten::lt(%15, %50) # <string>:103:27
  %17 : Float(*, *) = aten::type_as(%mask.1, %0) # <string>:104:26
  %18 : Float(*, *) = aten::mul(%17, %0) # <string>:104:26
  %res.1 : Float(*, *) = aten::mul(%18, %51) # <string>:104:26
  return (%res.1, %5, %6, %7, %49, %50, %mask.1, %50)
```

## Expected behavior

Previously, dropout node could be fused as:
```
graph(%x.1 : Float(*, *)):
  %17 : Float(*, *) = prim::DifferentiableGraph_0(%x.1)
  return (%17)
with prim::DifferentiableGraph_0 = graph(%0 : Float(*, *)):
  %50 : float = prim::Constant[value=0.90000000000000002]()
  %49 : bool = prim::Constant[value=1]()
  %7 : int = prim::Constant[value=1]() # <string>:103:64
  %6 : float = prim::Constant[value=1.]() # <string>:100:18
  %5 : None = prim::Constant()
  %res.1 : Float(*, *), %mask.1 : Bool(*, *) = prim::FusionGroup_0(%0)
  return (%res.1, %5, %6, %7, %49, %50, %mask.1, %50)
with prim::FusionGroup_0 = graph(%4 : Float(*, *)):
  %1 : float = prim::Constant[value=1.1111111111111112]()
  %9 : float = prim::Constant[value=0.90000000000000002]()
  %11 : int = prim::Constant[value=1]() # <string>:103:64
  %12 : Float(*, *) = aten::rand_like(%4, %11) # <string>:103:27
  %mask.1 : Bool(*, *) = aten::lt(%12, %9) # <string>:103:27
  %7 : Float(*, *) = aten::type_as(%mask.1, %4) # <string>:104:26
  %5 : Float(*, *) = aten::mul(%7, %4) # <string>:104:26
  %res.1 : Float(*, *) = aten::mul(%5, %1) # <string>:104:26
  return (%res.1, %mask.1)
```

## Additional context

I have checked the `FuseGraph` code path and I'm seeing the device inconsistency between inputs and outputs tensors for `rand_like` node. I'm suspecting it's related to changes in `shape_analysis.cpp` in: https://github.com/pytorch/pytorch/pull/33705/files

**Comment by ailzhang** [comment]
cc: @bhosmer for the PR mentioned in issue.

**Comment by eellison** [comment]
cc @Krovatkin 

We're trying to focus our efforts on profiling executor, which is the default on master & AFAIK will be the default for the next release. We'll accept bug fixes for the legacy executor but we're mostly working towards making the change. Shape analysis has proved to be a hassle to maintain and there isn't any plan on using it moving forward. The new profiling executor might introduce some behavioral changes that aren't compatible with the current shape inference, but we're hoping over the long term it will produce a better & more optimizable IR. 

We should probably stop recommending people switch to the legacy executor, since we don't plan to maintain it moving forward. Do you mind commenting with the issues you're having with the profiling executor that caused you to switch to legacy ? Ideally we'd just fix those.

**Comment by eellison** [comment]
I'm not that close to the source on this though so what I stated above may not reflect everyone's opinions 😛 

@Krovatkin thoughts ?

**Comment by Krovatkin** [comment]
As @eellison mentioned maintaining the legacy executor won't be our **top** priority moving forward. Please try using the new executor as much as possible and report any issues you run into with it. Those will be given a much higher priority.

We seem to internally agree we should fix **this** particular issue. We also welcome fixes from our contributors to the shape inference/legacy executor.

**Comment by Krovatkin** [comment]
cc: @zdevito @suo

**Comment by jjsjann123** [comment]
Thanks for the suggestion here. We fallback to legacy executor because of redundant profiling node does not get eliminated during optimization passes and was blocking the fusion. (I'll file a separate issue against that shortly)

We currently use legacy executor in many of our benchmark scripts to avoid regression.

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
